<a href="https://colab.research.google.com/github/ashishpal2702/Statistics/blob/master/hypothesis_testing_solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hypothesis Testing

The purpose of the test is to tell if there is any significant difference between two data sets.



## Overview

This module covers,

1) One sample and Two sample t-tests

2) ANOVA

3) Chi-Squared Tests

## Question 1 

*A student is trying to decide between two GPUs. He want to use the GPU for his research to run Deep learning algorithms, so the only thing he is concerned with is speed.*

*He picks a Deep Learning algorithm on a large data set and runs it on both GPUs 15 times, timing each run in hours. Results are given in the below lists GPU1 and GPU2.*

In [2]:
from scipy import stats 
import numpy as np

In [3]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU2 = np.array([11,13,10,13,12,9,11,12,12,11,12,12,10,11,13])

#Assumption: Both the datasets (GPU1 & GPU 2) are random, independent, parametric & normally distributed

In [6]:
len(GPU2)

15

Hint: You can import ttest function from scipy to perform t tests 

**First T test**

*One sample t-test*

Check if the mean of the GPU1 is equal to zero.
- Null Hypothesis is that mean is equal to zero.
- Alternate hypothesis is that it is not equal to zero.

# Answer

In [7]:
stats.ttest_1samp(GPU1,0)

Ttest_1sampResult(statistic=34.056241516158195, pvalue=7.228892044970457e-15)

Since, p < 0.05, so reject Null Hypothesis.

## Question 2

Given,

Null Hypothesis : There is no significant difference between data sets

Alternate Hypothesis : There is a significant difference

*Do two-sample testing and check whether to reject Null Hypothesis or not.*

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

# Answer

In [ ]:
stats.ttest_ind(GPU1,GPU2)

Ttest_indResult(statistic=-2.627629513471839, pvalue=0.013794282041452725)

p < 0.05 , reject Null Hypothesis.

## Question 3

He is trying a third GPU - GPU3.

In [ ]:
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])

#Assumption: Both the datasets (GPU1 & GPU 3) are random, independent, parametric & normally distributed

*Do two-sample testing and check whether there is significant differene between speeds of two GPUs GPU1 and GPU3.*

#### Answer:

In [ ]:
stats.ttest_ind(GPU1,GPU3)

Ttest_indResult(statistic=-1.4988943759093303, pvalue=0.14509210993138993)

Since, p > 0.05, so we can't reject Null Hypothesis. Hence, Mean of both GPU's are same. 

In [ ]:
stats.ttest_ind(GPU2,GPU3)

Ttest_indResult(statistic=0.814345071045956, pvalue=0.4223194951887398)

Since, p > 0.05, so we can't reject Null Hypothesis. Hence, Mean of both GPU's are same. 

## ANOVA

## Question 4 

If you need to compare more than two data sets at a time, an ANOVA is your best bet. 

*The results from three experiments with overlapping 95% confidence intervals are given below, and we want to confirm that the results for all three experiments are not significantly different.*

But before conducting ANOVA, test equality of variances (using Levene's test) is satisfied or not. If not, then mention that we cannot depend on the result of ANOVA

In [ ]:
import numpy as np

e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

#Assumption: All the 3 datasets (e1,e2 & e3) are random, independent, parametric & normally distributed

Perform levene test on the data

The Levene test tests the null hypothesis that all input samples are from populations with equal variances. Levene’s test is an alternative to Bartlett’s test bartlett in the case where there are significant deviations from normality.

source: scipy.org

For Null Hypothesis: Variance is equal
For Alternate Hypothesis: Variance is not equal

#### Answer:

In [ ]:
stats.levene(e1,e2,e3, center= 'median')

LeveneResult(statistic=2.6741725711150446, pvalue=0.12259792666001798)

Since, p > 0.05, so we accept Null Hypothesis. Hence, e1,e2 and e3 pass equality test of variance.

## Question 5

The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.

use stats.f_oneway() module to perform one-way ANOVA test

# Answer

In [ ]:
stats.f_oneway(e1,e2,e3)

F_onewayResult(statistic=2.51357622845924, pvalue=0.13574644501798466)

Since, p > 0.05, Fail to reject the Null Hypothesis. 

## Question 6 

You are a manager of a chinese restaurant. You want to determine whether the waiting time to place an order has changed in the past month from its previous population mean value of 4.5 minutes. 
State the null and alternative hypothesis.



# Answer

Null Hypothesis: Mean = 4.5

Alternate Hypothesis: Mean is not equal to 4.5

## Chi square test

## Question 7

Let's create a small dataset for dice rolls of four players

In [ ]:
import numpy as np

d1 = [1, 6, 3, 4]
d2 = [2, 5, 1, 3]
d3 = [4, 2, 3, 1]
d4 = [3, 4, 1, 2]
d5 = [1, 6, 3, 5]
d6 = [3, 2, 2, 1]

dice = np.array([d1, d2, d3, d4, d5, d6])

dof = (r-1)(c-1)

run the test using SciPy Stats library

Depending on the test, we are generally looking for a threshold at either 0.05 or 0.01. Our test is significant (i.e. we reject the null hypothesis) if we get a p-value below our threshold.

For our purposes, we’ll use 0.01 as the threshold.

use stats.chi2_contingency() module 

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table

Print the following:

- chi2 stat
- p-value




# Answer

In [ ]:
stats.chi2_contingency(dice)

(11.445004959326388,
 0.720458335452983,
 15,
 array([[2.88235294, 5.14705882, 2.67647059, 3.29411765],
        [2.26470588, 4.04411765, 2.10294118, 2.58823529],
        [2.05882353, 3.67647059, 1.91176471, 2.35294118],
        [2.05882353, 3.67647059, 1.91176471, 2.35294118],
        [3.08823529, 5.51470588, 2.86764706, 3.52941176],
        [1.64705882, 2.94117647, 1.52941176, 1.88235294]]))

Since, chi2 stat = 11.445004959326388, p value = 0.720458335452983, 

Hence, p > 0.05, so we accept Null Hypothesis.

## Question 8

### Z-test

Get zscore on the above dice data using stats.zscore module from scipy. Convert zscore values to p-value and take mean of the array.

# Answer

In [ ]:
dice

array([[1, 6, 3, 4],
       [2, 5, 1, 3],
       [4, 2, 3, 1],
       [3, 4, 1, 2],
       [1, 6, 3, 5],
       [3, 2, 2, 1]])

In [ ]:
z = stats.zscore(dice)
z

array([[-1.20604538,  1.09454091,  0.92847669,  0.89442719],
       [-0.30151134,  0.4975186 , -1.29986737,  0.2236068 ],
       [ 1.50755672, -1.29354835,  0.92847669, -1.11803399],
       [ 0.60302269, -0.09950372, -1.29986737, -0.4472136 ],
       [-1.20604538,  1.09454091,  0.92847669,  1.56524758],
       [ 0.60302269, -1.29354835, -0.18569534, -1.11803399]])

In [ ]:
p_values = stats.norm.sf(abs(z))
p_values

array([[0.1139    , 0.13685891, 0.17658018, 0.18554668],
       [0.3815123 , 0.3094117 , 0.09682322, 0.41153164],
       [0.06583401, 0.09791074, 0.17658018, 0.13177624],
       [0.2732468 , 0.46036917, 0.09682322, 0.32736042],
       [0.1139    , 0.13685891, 0.17658018, 0.05876243],
       [0.2732468 , 0.09791074, 0.42634184, 0.13177624]])

In [ ]:
mean = p_values.mean()
mean

0.20239343794918715

## Question 9

A Paired sample t-test compares means from the same group at different times.

The basic two sample t-test is designed for testing differences between independent groups. 
In some cases, you might be interested in testing differences between samples of the same group at different points in time. 
We can conduct a paired t-test using the scipy function stats.ttest_rel(). 

In [ ]:
before= stats.norm.rvs(scale=30, loc=100, size=500) ## Creates a normal distribution with a mean value of 100 and std of 30
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=500)

Test whether a weight-loss drug works by checking the weights of the same group patients before and after treatment using above data.

# Answer

In [ ]:
stats.ttest_rel(before,after)

Ttest_relResult(statistic=6.0138427443843305, pvalue=3.504861416014017e-09)

Null Hypothesis : There is no change in weight after drug.

Alternate Hypothesis : There is significant change in weight after drug.

p < 0.05, there is significant change in weight after drug.